earth_enterprise/src/common/khMTTypes.h

```diff
@@ -148,9 +151,173 @@ class khMTSet
+    return map.empty();
+  }
+  khMTMap(const khMTMap&) {
+    // todo: this is only used for setting up an empty dirtymap by AssetD and AssetVersionD
+    // should probably really remove this
+  }
+};
+
+// WARNING: Use this with extreme caution
+// It protects some obvious problems but inheriting from vector publicly isn't great
+template <typename T>
+class MTVector : public std::vector<T> {
+  private:
+    mutable std::mutex mtx;
+
+    MTVector(const MTVector& a, const std::lock_guard<std::mutex> &) : Base(a) { }
+  public:
+    typedef std::vector<T> Base;
+    typedef typename Base::const_iterator const_iterator;
+    void push_back(const T& v) {
+      std::lock_guard<std::mutex> lock(mtx);
+      Base::push_back(v);
+    }
+    MTVector() : Base() { }
+    // forward to private copy constructor to protect vector from modification
+    MTVector(const MTVector& a) : MTVector(a, std::lock_guard<std::mutex>(a.mtx)) { }
+    const T& operator[] (const size_t idx) const {
+      std::lock_guard<std::mutex> lock(mtx);
+      return Base::operator[](idx); 
+    }
+
+    bool operator==(const MTVector& x) const {
+      std::lock_guard<std::mutex> lock(mtx);
+      return true;
+    }
+
+    bool operator==(const Base& x) const {
+      std::lock_guard<std::mutex> lock(mtx);
+      return true;
+    }
+
+    void clear(void) {
+      std::lock_guard<std::mutex> lock(mtx);
+      Base::clear();
+    }
+
+    void shrink_to_fit(void) {
+      std::lock_guard<std::mutex> lock(mtx);
+      Base::shrink_to_fit();
+    }
+
+    const_iterator begin(void) const {
+      return Base::begin();
+    }
+
+    const_iterator end(void) const {
+      return Base::end();
+    }
+
+    size_t size(void) const {
+      std::lock_guard<std::mutex> lock(mtx);
+      return Base::size();
+    }
+};
+
+class khTargetedLock {
+  private:
+    // no such thing as a targeted lock which doesn't have a target
+    khTargetedLock() = delete;
+    khTargetedLock(const khTargetedLock&) = delete;
+    typedef std::unordered_map<std::string, std::shared_ptr<std::mutex>> FileLocks;
+    
+    static std::mutex& globalMutex(void) {
+      static std::mutex classMutex;
+      return classMutex;
+    }
+    std::shared_ptr<std::mutex> theFileMutex;
+    const std::string& s;
+    static FileLocks& getFileLocks(void) {
+      static FileLocks file_locks;
+      return file_locks;
+    } 
+  public:
+    khTargetedLock(const std::string& filename) : s(filename) {
+      FileLocks& file_mutexes = getFileLocks();
+      {
+        std::lock_guard<std::mutex> global_lock(globalMutex());
+        if (file_mutexes.find(s) == file_mutexes.end()) {
+          file_mutexes[s] = std::make_shared<std::mutex>();
+        }
+        theFileMutex = file_mutexes[s];
+      }
+      theFileMutex->lock();
+    }
+    ~khTargetedLock(void) {
+      theFileMutex->unlock();
+      std::lock_guard<std::mutex> global_lock(globalMutex());
+      if (theFileMutex.use_count() == 2) {
```

Comment:
I've heard use_count is just a rough count - it's not guaranteed to be exactly correct.